exonum-java-binding-core/rust/src/utils/jni_cache.rs

```diff
@@ -0,0 +1,158 @@
+use jni::{
+    objects::{GlobalRef, JMethodID},
+    sys::{jint, JNI_VERSION_1_8},
+    JNIEnv, JavaVM,
+};
+use std::os::raw::c_void;
+
+static mut OBJECT_CACHED_GET_CLASS: Option<JMethodID> = None;
+static mut CLASS_CACHED_GET_NAME: Option<JMethodID> = None;
+static mut THROWABLE_CACHED_GET_MESSAGE: Option<JMethodID> = None;
+
+static mut UTA_CACHED_EXECUTE: Option<JMethodID> = None;
+static mut UTA_CACHED_INFO: Option<JMethodID> = None;
+static mut UTA_CACHED_VERIFY: Option<JMethodID> = None;
+
+static mut USA_CACHED_STATE_HASHES: Option<JMethodID> = None;
+static mut USA_CACHED_CONVERT_TRANSACTION: Option<JMethodID> = None;
+
+static mut CLASS_CACHED_ERROR: Option<GlobalRef> = None;
+static mut CLASS_CACHED_TEE: Option<GlobalRef> = None;
+
+#[allow(non_snake_case)]
+#[no_mangle]
+pub extern "system" fn JNI_OnLoad(vm: JavaVM, _: *mut c_void) -> jint {
+    let env = vm.get_env().expect("Cannot get reference to the JNIEnv");
+    cache_methods(&env);
+
+    JNI_VERSION_1_8
+}
+
+/// Caches all required classes and methods ids.
+pub fn cache_methods(env: &JNIEnv) {
+    unsafe {
+        OBJECT_CACHED_GET_CLASS =
```

Comment:
Is this implementation thread-safe? Is a write in the thread performing the initialization *guaranteed* to be visible in a thread performing a read?

Reply:
🤔 Actually, it's not thread-safe.